0015-ilp-addresses/0015-ilp-addresses.md

```diff
@@ -0,0 +1,158 @@
+
+1. The address MUST begin with a prefix indicating the allocation scheme. See [Allocation Schemes](#allocation-schemes) for more information.
+2. Each "segment" of the address MUST contain one or more of the following characters:
+    - Alphanumeric characters, upper or lower case. (Addresses are **case-sensitive** so that they can contain data encoded in formats such as base64url.)
+    - Underscore (`_`)
+    - Tilde (`~`)
+    - Hyphen (`-`)
+3. Each segment MUST be separated from other segments by a period character (`.`).
+4. Address prefixes MUST end in a period (`.`) character and MAY contain any number of segments after the allocation scheme prefix.
+6. Destination addresses MUST NOT end in a period (`.`) character, and MUST contain at least two segments after the allocation scheme prefix.
+
+The following ABNF specification defines the format for all ILP addresses and address prefixes:
+
+```abnf
+address     = scheme separator *prefix [segment]
+                    ; the last segment is REQUIRED for destination addresses
+
+scheme      = "g" / "private" / "example" / "peer" / "self" /
+              "test1" / "test2" / "test3"
+
+separator   = "."
+
+prefix      = 1*(segment separator)
+
+segment     = 1*( ALPHA / DIGIT / "_" / "~" / "-" )
+```
+
+You can also use the following regular expressions to verify the same requirements:
+
+| Address Type        | Regular Expression                                     |
+|:--------------------|:-------------------------------------------------------|
+| All addresses       | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*([a-zA-Z0-9_~-]+)?$` |
+| Address prefix      | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)*$` |
+| Destination address | `^(g|private|example|peer|self|test[1-3])\.([a-zA-Z0-9_~-]+\.)+[a-zA-Z0-9_~-]+$` |
+
+
+## Allocation Schemes
+
+The allocation scheme is the first part of an address, which indicates how the address is assigned. Here is a summary of the prefixes that are currently defined:
+
+| Prefix                       | Allocation Scheme             | Definition and Use Case |
+|:-----------------------------|:------------------------------|:--------------|
+| `g.`                         | [General Allocation Scheme][] | Most ILP addresses used to send and receive real money. |
+| `private.`                   | Private allocation            | For ILP addresses that only have meaning in a private subnet or intranet. Analogous to the [192.168.0.0/16 range in IPv4](https://en.wikipedia.org/wiki/Private_network). |
+| `example.`                   | Examples                      | For "non-real" addresses that are used as examples or in documentation. Analogous to ["555 phone numbers"](https://en.wikipedia.org/wiki/555_%28telephone_number%29) in the USA. |
+| `test1.`, `test2.`, `test3.` | Testing                       | For addresses used in tests, such as unit or integration tests of compatible software. |
```

Comment:
Well, I picked `test1`, `test2`, `test3` for three reasons:

1. As a nod to [RFC 5737](https://tools.ietf.org/html/rfc5737).
2. Because our [integration tests]() currently use `test1` and `test2`.
3. Because people like when things come in threes.

Argument #2 is probably the strongest - you should always try to write your spec such that it describes what you already do, unless you have a strong reason to change it.